pysys/utils/threadutils.py

```diff
@@ -0,0 +1,188 @@
+		self.stopping.set()
+		return self
+		
+	def join(self, timeout=None, abortOnError=False):
+		"""
+		Wait until this thread terminates, and adds a TIMEDOUT or BLOCKED 
+		outcome to the owner test if it times out or raises an exception.
+		
+		If you wish to request the thread to terminate rather than waiting for 
+		it to reach the end of its target function on its own, call L{stop} 
+		before joining the thread. 
+		
+		If a join times out, the thread is automatically requested to stop 
+		as soon as possible. 
+		
+		Note that if the thread raises an Exception after it was requested to 
+		stop this is logged but does not result in a failure outcome, 
+		since failures during cleanup are usually to be expected. 
+		
+		@param timeout: The time in seconds to wait. Usually this should be 
+		left at the default value of None which uses a default timeout 
+		of L{constants.TIMEOUTS}C{['WaitForProcessStop']}. 
+		Note that unlike Python's `Thread.join` method, infinite timeouts 
+		are not supported. 
+		
+		@param abortOnError: Set to True if you wish this method to 
+		immediately abort with an exception if the background thread times out 
+		or raises an Exception. The default is False, which adds the failure 
+		outcome but does not raise an exception. 
+		"""
+		outcomereported = self.__outcomeReported
+		self.__outcomeReported = True # only do this once
+		
+		if not timeout: timeout = self.joinTimeoutSecs
+		assert self.joinTimeoutSecs > 0, self.joinTimeoutSecs
+
+		if self.thread.isAlive() or (not outcomereported):
+			# only log it the first time
+			self.log.info('Joining background thread %s'%self)
+		starttime = time.time()
+		
+		# don't call thread.join for the entire time, since on windows that 
+		# leaves no opportunity to detect keyboard interrupts
+		if self.__kbdrInterrupt: raise self.__kbdrInterrupt # avoid repeatedly joining same thread
+		while self.thread.isAlive() and time.time()-starttime < timeout:
+			try:
+				self.thread.join(1)
+			except KeyboardInterrupt as ex: # progra: no cover
+				self.__kbdrInterrupt = ex
+				raise
+		
+		if self.thread.isAlive():
+		
+			if not outcomereported:
+				self.owner.addOutcome(TIMEDOUT, 'Background thread %s is still running after waiting for allocated timeout period (%d secs)'%(
+					self, timeout), abortOnError=abortOnError)
+				try:
+					self.log.warning('Stack of hanging thread %s: \n%s', str(self), 
+						''.join(traceback.format_stack(sys._current_frames()[self.thread.ident])))
+				except Exception as ex:
+					self.log.debug('Failed to get stack of hanging thread: %s: %s', ex.__class__.__name__, ex)
+			self.stop() # ensure it stops as quickly as possible
+
+		elif self.exception is not None:
+			if not outcomereported:
+				self.owner.addOutcome(BLOCKED, 'Background thread %s failed with %s: %s'%(
+					self, self.exception.__class__.__name__, self.exception), abortOnError=abortOnError)
+		elif time.time()-starttime>10: # alert user only if it took a long time
```

Comment:
i guess this is a subjective one. it's not a function of the timeout, and unlike a tiemout it's not the kind of number you'd want to change so I don't think a constant is helpful. i like to see in the test output when an operation has taken a non-trivial (=10s) amount of time ; we could print it in all cases but then you end up cluttering the log with lots of joins that tooks 0 seconds which isn't helpful. so I think this is ok

Reply:
Definitely subjective.  I was imagining a case where I expect a thread to take, say, 20s to join (because it has to shut down a connection to some slow remote service or something) where I will end up seeing this message every time even though I set the timeout to a suitably large value.  So you _could_ say that you only want to know if the join took > 75% of the timeout... I know, I just introduced another arbitrary constant there :-)  And fixed timeouts of any kind are one of my pet annoyances, as you know.

Anyway, I concede that making it work the way I want it to is probably more effort than it is really worth, so I'm happy to leave it as it is for now.

Reply:
Yes I'd be 100% with you for anything that's really a timeout, would always have that configurable; but for things like this which are really optional information to help the user (with a tradeoff between useful vs cluttered) I lean towards a pramatic and heuristic approach (and extra config defo feels like overkill). Anyway, always good to discuss these things.